### PR TITLE
Update batch norm to compute forward grads for saved_mean and saved_var when input requires grad

### PR DESCRIPTION
**Pull request by soulitzer**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #81293

This change lets us get rid of the additional decomposition table specifically used for jvps

I'm actually not completely sure we want to land this. Some downsides of the current approach:
- The requires_grad-ness of the return lies about the backward differentiability of saved_mean and saved_var!
   - a simple mitigation: add an assert when non-zero gradient is passed as grad_saved_mean/var
   - how bad that is depends on how "user-facing" we consider this API
- The decomposition and actual function diverges (we could either skip the test or update actual function also return a zero-filled tensor)

Some alternatives:
- Actually tell the truth about differentiability - update codegen to allow functions forward differentiable, but not backward differentiable functions
   - This is not an entirely new concept actually because differentiability is not bidirectional in the case of detach
- Actually make the gradients flow properly
   - This is the cleaner solution, but requires a rewrite of the double backward formula.
- Continue with the status quo:
  - We'd have to stick with the current hack, but maybe it isn't so bad considering the alternatives.

This PR:
- Gets rid of the additional decomposition table
- Adds a bunch of random changes to the decomposition necessary to get it to JIT compile. I chose not to use the decomposition from before because it is missing subtle things like casting to the right intermediate dtype (I'm not sure if there are other differences).

### CODE REVIEW

File: torch/testing/_internal/common_methods_invocations.py [hunk 1]
```diff
@@ -1849,7 +1849,7 @@ def sample_inputs_batch_norm(op_info, device, dtype, requires_grad, **kwargs):
         ((0,), {'training': False}),
         ((3, 2, 3, 4), {'training': True, 'momentum': -1.0, 'eps': 0.5}),
         ((3, 2, 3, 4), {'training': False, 'momentum': -1.0, 'eps': 0.5}),
-        ((2, 1), {}),
+        ((2, 2), {}),
```
Comment (soulitzer) [inline torch/testing/_internal/common_methods_invocations.py:1852]: What we really need to do here is increase the tolerance, the tests actually on most random seeds

File: tools/autograd/derivatives.yaml [hunk 2]
```diff
@@ -1130,13 +1130,13 @@
...
-  result0: batch_norm_jvp(input_p, input_t, weight_p, weight_t, bias_p, bias_t, running_mean, running_var, result1, result2, training, eps)
+  input, weight, bias: "grads[0].defined() ? native_batch_norm_backward(grads[0], input, weight, running_mean, running_var, result1, result2, training, eps, grad_input_mask) : std::tuple<Tensor, Tensor, Tensor>()"
+  result0, result1, result2: batch_norm_jvp(input_p, input_t, weight_p, weight_t, bias_p, bias_t, running_mean, running_var, result1, result2, training, eps)
 
 - name: native_batch_norm_backward(Tensor grad_out, Tensor input, Tensor? weight, Tensor? running_mean, Tensor? running_var, Tensor? save_mean, Tensor? save_invstd, bool train, float eps, bool[3] output_mask) -> (Tensor, Tensor, Tensor)
   input, weight, grad_out: batchnorm_double_backward(input, weight, grads[0], grads[1], grads[2], grad_out, running_mean, running_var, train, eps, save_mean, save_invstd, grad_input_mask)
-  save_mean: not_implemented("native_batch_norm_backward save_mean")
-  save_invstd: not_implemented("native_batch_norm_backward save_invstd")
+  save_mean: Tensor()
```
[the remark this answers is not in the document]
Comment (soulitzer) [reply tools/autograd/derivatives.yaml:1143]: Yeah they are the right gradients - I had to think for a while to realize this, but what is happening is that because save_mean and save_invstd can be both written in terms of input. batchnorm_double_backward computes the backward of batchnorm_backward as if the entire computation were computed in terms of inputs (as save_mean and save_invstd are soley used to reduce recomputation). I will update this PR to add a note for this.
